Add 0036-trusted-artifacts

lcarva opened the pull request.

lcarva commented:
@ralphbean, @brianwcook, @arewm PTAL :pray:

zregvart commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,47 @@
+## Context
+
+One of the properties of Konflux is that users should be allowed to include their own Tekton Tasks
+in a build Pipeline, e.g. to execute unit tests, without invalidating the build process. This is
```
Perhaps

```suggestion
in a build Pipeline, e.g. to execute unit tests, without jeopardizing the integrity of the build process. This is
```

lcarva replied:
Better. Thanks.

zregvart commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,47 @@
+in a build Pipeline, e.g. to execute unit tests, without invalidating the build process. This is
+distinct from other build systems where a rigid process prevents users from applying such
+customizations. To support this, Konflux build Pipelines use Trusted Artifacts to securely share
+data between Tasks. Enterprise Contract is then responsible for verifying that Trusted Artifacts
```
```suggestion
files between Tasks. Enterprise Contract is then responsible for verifying that Trusted Artifacts
```

lcarva replied:
Good to be specific. Also modified other occurrences.

lcarva commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,47 @@
+were properly used in the parts of the build Pipeline that affect the build outcome, e.g.
+`git-clone` and `buildah` Tasks.
+
+Trusted Artifacts is inspired by the upcoming work being done by the Tekton Community,
```
The upstream work will require changes to the Tekton resources IIUC. I don't think there's a way to avoid a migration. But hopefully, this is isolated to Task definitions and we can switch over without requiring a migration from the user side.

ifireball commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,47 @@
+[TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md). The
+Konflux version is meant to be a stop-gap until that feature is implemented and ready to be used.
+When the time comes, the Konflux implementation should align with what is provided by the Tekton
+Community and this ADR may need a revision.
```
A brief summary of what make trusted artifacts "trusted" and how do they enable pipeline security would be welcome here.

All I could gather from this text is that they are stored in an OCI registry...

lcarva replied:
Excellent idea! Added a paragraph. Let me know what you think

ifireball replied:
looks good, what is preventing tampering in the case? The fact that the checksum is passed via channel that is internal to the pipeline?

ifireball commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,53 @@
+location of the archive and its checksum digests are recorded as a Task result. The process of
+*consuming* a Trusted Artifact extracts such archive, while verying its checksum digest, into a
+volume only accessible to the Task, e.g. `emptyDir`. The name and checksum digest of the archive is
+provided via Task parameters.
```
```suggestion
provided via Task parameters.

With the artifact checksums being passed over the task parameters we can ensure that the artifacts are bot being tampered with as they are passed between different pipeline tasks.
```

Trying to summarize our other thread

lcarva replied:
TY! I reworded this a bit. Let me know what you think

simonbaird commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,54 @@
+
+In brief, the processes of *creating* a Trusted Artifact wraps files into an archive. Then, the
+location of the archive and its checksum digests are recorded as a Task result. The process of
+*consuming* a Trusted Artifact extracts such archive, while verifying its checksum digest, into a
```
```suggestion
*consuming* a Trusted Artifact extracts such an archive, while verifying its checksum digest, into a
```

arewm commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,54 @@
+process. This is distinct from other build systems where a rigid process prevents users from
+applying such customizations. To support this, Konflux build Pipelines use Trusted Artifacts to
+securely share files between Tasks. Enterprise Contract is then responsible for verifying that
+Trusted Artifacts were properly used in the parts of the build Pipeline that affect the build
```
It also is used to verify that the contents were not tainted by untrusted tasks.

lcarva replied:
I think this falls under the "Trusted Artifacts were properly used" comment.

arewm commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,54 @@
+Trusted Artifacts is inspired by the upcoming work being done by the Tekton Community,
+[TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md). The
+Konflux version is meant to be a stop-gap until that feature is implemented and ready to be used.
+When the time comes, the Konflux implementation should align with what is provided by the Tekton
+Community and this ADR may need a revision.
```
Can we state that a new ADR should be created to represent the fact that we are changing as well?

lcarva replied:
Good idea. Done.

arewm commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,54 @@
+When the time comes, the Konflux implementation should align with what is provided by the Tekton
+Community and this ADR may need a revision.
+
+In brief, the processes of *creating* a Trusted Artifact wraps files into an archive. Then, the
```
Do we need to be more specific about what the archive is? These archives also need to be tamper-proof.

lcarva replied:
I really don't want this ADR to be a design doc of Trusted Artifacts. I think that belongs somewhere else. https://github.com/konflux-ci/build-trusted-artifacts is likely a good place for it.

arewm commented on ADR/0036-trusted-artifacts.md:
```diff
@@ -0,0 +1,54 @@
+
+## Decision
+
+Sharing files between Tasks is done via Trusted Artifacts backed by OCI storage.
```
Do we want to state that we cannot use the referrer's api here because these archives are generated before we produce an image? I don't know if anyone might wonder about it. (no is a fine response).

lcarva replied:
I think that's too low level for this doc.

arewm reviewed (approved):
The comments are all relatively minor. Thanks for writing this up!

lcarva commented:
@arewm, I believe I addressed all of your comments. Let me know what you think.

(BTW, I don't have merge permission here, so if you're happy with it, please merge :pray:)
